server/database/challenges.ts

```diff
@@ -36,4 +45,19 @@ export const upsertChallenge = async ({ id, data }: DatabaseChallenge): Promise<
     `,
   [id, data]
   )
+  if (data.tags instanceof Array) {
+    const existingTags = await getAllTagsByChallenge({ challid: id })
+    for (const existingTag of existingTags) {
+      if (Array.isArray(data.tags)) {
+        // At this point, data.tags should be an array of Tags, and if not, user error
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+        if (!data.tags.some((t) => { return t.name === existingTag.name && t.metatag === existingTag.metatag })) {
+          await removeTagByName({ ...existingTag, challid: id })
+        }
+      }
+    }
+    for await (const tag of data.tags) {
+      await setTag({ ...tag, challid: id })
+    }
+  }
```

Comment:
This whole part is non-atomic but it's probably fine because challenge updates are infrequent

Reply:
I'll put it in a transaction regardless, can't hurt